flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/RetryingExecutor.java

```diff
@@ -172,16 +225,19 @@ private RetriableTask next() {
                     actionCompleted,
                     runnable,
                     retryPolicy,
-                    executorService,
-                    attemptsPerTaskHistogram);
+                    blockingExecutor,
+                    attemptsPerTaskHistogram,
+                    scheduler,
+                    failureCallback,
+                    activeAttempts);
         }
 
         private Optional<ScheduledFuture<?>> scheduleTimeout() {
             long timeout = retryPolicy.timeoutFor(current);
             return timeout <= 0
                     ? Optional.empty()
                     : Optional.of(
-                            executorService.schedule(
+                            scheduler.schedule(
```

Comment:
I think it's the IO, not the timer thread pool that is exhausted. The test with `maxAttempts=2`  succeeds with `nThreads=2` passed to `RetryingExecutor`.
This happens because `BlockingUploader` in the test waits indefinitely (and is not interrupted).

Not interrupting is intentional here - the upload should be completed by an attempt that finishes the first (it can be "timed out" but can still complete before subsequent attempts).

Reply:
Thanks for your explaination.

I did not get the idea that why we does not interrupt the timeout attempt in test? I can see the tasks would be failed in `BatchingStateChangeUploader` if timedout.

Reply:
After discussed offline, the confusing problem is that current implementation does not discard unnecessarily uploaded state.

Reply:
Right, I've opened #18976 to address this.